README.md
# Number Humanizer

A ruby gem to convert numbers into words. It has been designed to support multiple languages and currently support only Arabic and English languages.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'number_humanizer'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install number_humanizer

## Usage

You have two ways to use the gem:

1. Using service class

```ruby
require 'number_humanizer'

NumberHumanizer::Manager.new(1500).call.result
#=> "ألف و خمسة مائة"
```

2. Call monkey patching methods added to Integer and Float classes
```ruby
require 'number_humanizer'

1500.to_word
#=> "ألف و خمسة مائة"
```

Also you can pass currency option as follows:

```ruby
require 'number_humanizer'

NumberHumanizer::Manager.new(1500, currency: 'ريال').call.result
#=> "ألف و خمسة مائة ريال"

1500.to_word(currency: 'ريال')
#=> "ألف و خمسة مائة ريال"
```

Note that sub currency is also supported but not fully functional so if you need to use it then you can pass amount and sub amount separately to the service and you need to combine.

## Configurations
You have to way for setting configurations in number humanizer gem:
- Set configurations globally: this will be used in all conversions. You can add the following snippet in initializers folder or any file that is loaded automatically. Note that this configuration can be changed at any time by calling this code.

```ruby
# initializers/number_humanizer.rb
require 'number_humanizer'

NumberHumanizer.setup do |config|
  # default language is arabic
  config.language = :arabic

  # currency is not set by default and you can set from here
  config.currency = 'ريال'

  # all upcoming configs will be placed here
end
```

- Pass configuratoins instantly whether when using service object or monkey patched methods so that global configurations will be ignored

```ruby
require 'number_humanizer'

NumberHumanizer::Manager.new(1500, language: :arabic, currency: 'ريال').call.result
#=> "ألف و خمسة مائة ريال"

1500.to_word(language: :arabic, currency: 'ريال')
#=> "ألف و خمسة مائة ريال"
```

## Supported Languages
Initially, gem was supporting Arabic language only but now English language is supported too. Here are the details needed for each language:

- **Arabic Language**

  This is the default one and you don't need to specify it but if you set other language and want to have arabic word for given number then you can pass it to the service or as args in `to_word` monkey patched method as explained in configurations section.

  *Note that accepted values are: `arabic` and `ar`*.

- **English Language**

  Like arabic language, english language can be configured globally or instantly as described in configuration section. Here you just need to set `language` config.

  *Note that accepted values are: `english` and `en`*.


## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/abahaggag/number_humanizer. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the NumberHumanizer project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/number_humanizer/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [0.1.0] - 2022-05-28
  - Release the gem with basic functionalities


## [0.1.1] - 2022-06-03
  - Bug fixes and specs added


## [0.2.0] - 2022-06-04
  - Enhance and refactor languages structure to support multiple languages.
  - Add English language humanizer and its specs.
  - Setup configurations structure and basic configs.
  - Fix issues.

spec/spec_helper.rb
require "bundler/setup"
require "number_humanizer"
require "byebug"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/number_humanizer_spec.rb
RSpec.describe NumberHumanizer do
  it "has a version number" do
    expect(NumberHumanizer::VERSION).not_to be nil
  end

  context 'when arabic language is configured' do
    before do
      NumberHumanizer.setup do |config|
        config.language = :arabic
      end
    end

    it 'converts basic numbers correctly' do
      expect(0.to_word).to eq('صفر')
      expect(1.to_word).to eq('واحد')
      expect(2.to_word).to eq('إثنان')
      expect(3.to_word).to eq('ثلاثة')
      expect(4.to_word).to eq('أربعة')
      expect(5.to_word).to eq('خمسة')
      expect(6.to_word).to eq('ستة')
      expect(7.to_word).to eq('سبعة')
      expect(8.to_word).to eq('ثمانية')
      expect(9.to_word).to eq('تسعة')
      expect(10.to_word).to eq('عشرة')
      expect(20.to_word).to eq('عشرون')
      expect(30.to_word).to eq('ثلاثون')
      expect(40.to_word).to eq('أربعون')
      expect(50.to_word).to eq('خمسون')
      expect(60.to_word).to eq('ستون')
      expect(70.to_word).to eq('سبعون')
      expect(80.to_word).to eq('ثمانون')
      expect(90.to_word).to eq('تسعون')
      expect(100.to_word).to eq('مائة')
      expect(200.to_word).to eq('مئتان')
      expect(1_000.to_word).to eq('ألف')
      expect(2_000.to_word).to eq('ألفان')
      expect(1_000_000.to_word).to eq('مليون')
      expect(1_000_000_000.to_word).to eq('مليار')
      expect(1_000_000_000_000.to_word).to eq('بليون')
      expect(1_000_000_000_000_000.to_word).to eq('بليار')
      expect(1_000_000_000_000_000_000.to_word).to eq('تريليون')
      expect(1_000_000_000_000_000_000_000.to_word).to eq('تريليار')
      expect(1_000_000_000_000_000_000_000_000.to_word).to eq('كريليون')
      expect(1_000_000_000_000_000_000_000_000_000.to_word).to eq('كريليار')
      expect(1_000_000_000_000_000_000_000_000_000_000.to_word).to eq('سنكليون')
      expect(1_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('سنكليار')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('سيزيليون')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('سيزيليار')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('سيتليون')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('سيتليار')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('ويتليون')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('ويتليار')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('تيفليون')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('تيفليار')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('ديشليون')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('ديشلي')
    end

    it 'converts numbers with combined segments correctly' do
      expect(11.to_word).to eq('احد عشر')
      expect(12.to_word).to eq('اثنا عشر')
      expect(13.to_word).to eq('ثلاثة عشر')
      expect(14.to_word).to eq('أربعة عشر')
      expect(15.to_word).to eq('خمسة عشر')
      expect(16.to_word).to eq('ستة عشر')
      expect(17.to_word).to eq('سبعة عشر')
      expect(18.to_word).to eq('ثمانية عشر')
      expect(19.to_word).to eq('تسعة عشر')

      expect(101.to_word).to eq('مائة و واحد')
      expect(102.to_word).to eq('مائة و إثنان')
      expect(110.to_word).to eq('مائة و عشرة')
      expect(201.to_word).to eq('مئتان و واحد')
      expect(202.to_word).to eq('مئتان و إثنان')
      expect(230.to_word).to eq('مئتان و ثلاثون')
      expect(301.to_word).to eq('ثلاثة مائة و واحد')
      expect(302.to_word).to eq('ثلاثة مائة و إثنان')
      expect(350.to_word).to eq('ثلاثة مائة و خمسون')
      expect(999.to_word).to eq('تسعة مائة و تسعة و تسعون')
      expect(1_001.to_word).to eq('ألف و واحد')
      expect(1_002.to_word).to eq('ألف و إثنان')
      expect(2_001.to_word).to eq('إثنان ألف و واحد')
      expect(2_002.to_word).to eq('إثنان ألف و إثنان')
      expect(17_006.to_word).to eq('سبعة عشر ألف و ستة')
      expect(201_004.to_word).to eq('مئتان و واحد ألف و أربعة')
      expect(1_000_400.to_word).to eq('مليون و أربعة مائة')
      expect(2_000_345_003.to_word).to eq('إثنان مليار و ثلاثة مائة و خمسة و أربعون ألف و ثلاثة')
      expect(99_100_000_020.to_word).to eq('تسعة و تسعون مليار و مائة مليون و عشرون')
    end
  end

  context 'when english language is configured' do
    before do
      NumberHumanizer.setup do |config|
        config.language = :english
      end
    end

    it 'converts basic numbers correctly' do
      expect(0.to_word).to eq('zero')
      expect(1.to_word).to eq('one')
      expect(2.to_word).to eq('two')
      expect(3.to_word).to eq('three')
      expect(4.to_word).to eq('four')
      expect(5.to_word).to eq('five')
      expect(6.to_word).to eq('six')
      expect(7.to_word).to eq('seven')
      expect(8.to_word).to eq('eight')
      expect(9.to_word).to eq('nine')
      expect(10.to_word).to eq('ten')
      expect(20.to_word).to eq('twenty')
      expect(30.to_word).to eq('thirty')
      expect(40.to_word).to eq('forty')
      expect(50.to_word).to eq('fifty')
      expect(60.to_word).to eq('sixty')
      expect(70.to_word).to eq('seventy')
      expect(80.to_word).to eq('eighty')
      expect(90.to_word).to eq('ninety')
      expect(100.to_word).to eq('hundred')
      expect(200.to_word).to eq('two hundred')
      expect(1_000.to_word).to eq('thousand')
      expect(2_000.to_word).to eq('two thousand')
      expect(1_000_000.to_word).to eq('million')
      expect(1_000_000_000.to_word).to eq('billion')
      expect(1_000_000_000_000.to_word).to eq('trillion')
      expect(1_000_000_000_000_000.to_word).to eq('quadrillion')
      expect(1_000_000_000_000_000_000.to_word).to eq('quintillion')
      expect(1_000_000_000_000_000_000_000.to_word).to eq('sextillion')
      expect(1_000_000_000_000_000_000_000_000.to_word).to eq('septillion')
      expect(1_000_000_000_000_000_000_000_000_000.to_word).to eq('octillion')
      expect(1_000_000_000_000_000_000_000_000_000_000.to_word).to eq('nonillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('decillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('undecillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('duodecillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('tredecillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('quattuordecillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('quindecillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('sexdecillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('septemdecillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('octodecillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('novemdecillion')
      expect(1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000.to_word).to eq('vigintillion')
    end

    it 'converts numbers with combined segments correctly' do
      expect(11.to_word).to eq('eleven')
      expect(12.to_word).to eq('twelve')
      expect(13.to_word).to eq('thirteen')
      expect(14.to_word).to eq('fourteen')
      expect(15.to_word).to eq('fifteen')
      expect(16.to_word).to eq('sixteen')
      expect(17.to_word).to eq('seventeen')
      expect(18.to_word).to eq('eighteen')
      expect(19.to_word).to eq('nineteen')

      expect(99.to_word).to eq('ninety nine')
      expect(101.to_word).to eq('hundred and one')
      expect(102.to_word).to eq('hundred and two')
      expect(110.to_word).to eq('hundred and ten')
      expect(201.to_word).to eq('two hundred and one')
      expect(202.to_word).to eq('two hundred and two')
      expect(230.to_word).to eq('two hundred and thirty')
      expect(301.to_word).to eq('three hundred and one')
      expect(302.to_word).to eq('three hundred and two')
      expect(350.to_word).to eq('three hundred and fifty')
      expect(999.to_word).to eq('nine hundred and ninety nine')
      expect(1_001.to_word).to eq('thousand and one')
      expect(1_002.to_word).to eq('thousand and two')
      expect(2_001.to_word).to eq('two thousand and one')
      expect(2_002.to_word).to eq('two thousand and two')
      expect(17_006.to_word).to eq('seventeen thousand and six')
      expect(66_155.to_word).to eq('sixty six thousand and hundred and fifty five')
      expect(201_004.to_word).to eq('two hundred and one thousand and four')
      expect(1_000_400.to_word).to eq('million and four hundred')
      expect(2_000_345_003.to_word).to eq('two billion and three hundred and forty five thousand and three')
      expect(99_100_000_020.to_word).to eq('ninety nine billion and hundred million and twenty')
    end
  end
end


lib/number_humanizer.rb
require "number_humanizer/version"
require "number_humanizer/monkey_patchers"
require "number_humanizer/configuration"
require "number_humanizer/languages/base"
require "number_humanizer/languages/arabic"

module NumberHumanizer
  extend NumberHumanizer::Configuration.new(
    language: :arabic,
    currency: nil,
    sub_currency: nil
  )

  def self.setup
    yield self
  end

  class LanguageNotSupportedError < StandardError
  end

  class Manager
    attr_reader :result, :number, :language, :args

    def initialize(number, **args)
      @number = number
      @language = args[:language] || NumberHumanizer.language
      @args = args
    end

    def call
      @result = language_service_class.new(number, **args).call.result
      self
    end

    private

    def language_service_class
      @language_service_class ||= \
        case language.downcase.to_sym
        when :arabic, :ar then NumberHumanizer::Languages::Arabic
        else raise(LanguageNotSupportedError, "Language specified (#{language}) not supported")
        end
    end
  end
end


lib/number_humanizer/languages/base.rb
module NumberHumanizer::Languages
  class Base
    attr_reader :result

    def initialize(number, **args)
      @number = number
      @real_part, @fraction_part = exctract_real_and_fraction_parts
      @currency = args[:currency] || NumberHumanizer.currency
      @sub_currency = args[:sub_currency] || NumberHumanizer.sub_currency
    end

    def call
      @result = process_result
      self
    end

    private

    attr_reader :number, :real_part, :fraction_part, :currency, :sub_currency,
                :processed_real_part, :processed_fraction_part

    def exctract_real_and_fraction_parts
      number.abs.to_s.split('.').map(&:to_i)
    end

    def process_result
      @processed_real_part = process_real_part
      @processed_fraction_part = process_fraction_part
      final_result = [
        [processed_real_part, currency].compact.join(' '),
      ]
      final_result << [processed_fraction_part, sub_currency].compact.join(' ') if processed_fraction_part.present?
      final_result.compact.join(decimal_point_connector).insert(0, process_minus)
    end

    def unit(unit_key); self.class::UNITS_DICTIONARY[unit_key]; end
    def word_lookup(segment); self.class::NUMBERS_DICTIONARY[segment]; end
    def and_connector; self.class::AND_CONNECTOR; end
    def decimal_point_connector; self.class::DECIMAL_POINT_CONNECTOR; end
    def minus_word; self.class::Minus_word; end

    def process_minus
      return minus_word if number.negative?
      ''
    end

    def process_real_part
      number_word = word_lookup(real_part)
      return number_word if number_word.present?
      process_number(real_part)
    end

    def process_fraction_part
      return if fraction_part.nil? || fraction_part.zero?
      fraction_part
      # process_number(fraction_part)
    end

    def process_number(number)
      segments = get_number_segments(number)
      segments_size = segments.size
      "#{segments.map!.with_index { |segment, i| process_segment(segment, unit_key: segments_size - i) }.compact.reject{ |sw| sw == '' }.join(and_connector)}"
    end

    def get_number_segments(number)
      return [0] if number.zero?
      segments = []
      while(number != 0)
        segments.insert(0, number % 1000)
        number /= 1000
      end
      segments
    end

    def process_segment(segment, unit_key:)
      word = "#{segment_to_word(segment, unit_key)}"
      word << " #{unit(unit_key)}" if unit_key != 1 && segment > 0
      word.squish!
    end

    def segment_to_word(segment, unit_key)
      return '' if segment.zero? || (segment == 1 && unit_key != 1)
      return word_lookup(segment) if word_lookup(segment).present?
      return process_words_for_tens(segment) if segment < 100
      return process_words_for_hundreds(segment) if segment < 1000
    end

    def process_words_for_tens(segment)
      return word_lookup(segment) if word_lookup(segment).present?
      "#{word_lookup(segment % 10)}#{segment / 10 == 1 ? ' ' : and_connector}#{word_lookup(segment / 10 * 10)}"
    end

    def process_words_for_hundreds(segment)
      word = "#{process_hundred_part(segment)}"
      word << "#{and_connector}#{process_words_for_tens(segment % 100)}" if segment % 100 > 0
      word
    end

    def process_hundred_part(segment)
      hundred_digit = segment / 100
      return '' if hundred_digit.zero?
      (1..2).cover?(hundred_digit) ? word_lookup(hundred_digit * 100) : "#{word_lookup(hundred_digit)} #{word_lookup(100)}"
    end
  end
end


lib/number_humanizer/languages/arabic.rb
module NumberHumanizer::Languages
  class Arabic < Base
    UNITS_DICTIONARY = {
      1 => '',
      2 => 'ألف',
      3 => 'مليون',
      4 => 'مليار',
      5 => 'بليون',
      6 => 'بليار',
      7 => 'تريليون',
      8 => 'تريليار',
      9 => 'كريليون',
      10 => 'كريليار',
      11 => 'سنكليون',
      12 => 'سنكليار',
      13 => 'سيزيليون',
      14 => 'سيزيليار',
      15 => 'سيتليون',
      16 => 'سيتليار',
      17 => 'ويتليون',
      18 => 'ويتليار',
      19 => 'تيفليون',
      20 => 'تيفليار',
      21 => 'ديشليون',
      22 => 'ديشلي'
    }.freeze

    NUMBERS_DICTIONARY = {
      0   => 'صفر',
      1   => 'واحد',
      2   => 'إثنان',
      3   => 'ثلاثة',
      4   => 'أربعة',
      5   => 'خمسة',
      6   => 'ستة',
      7   => 'سبعة',
      8   => 'ثمانية',
      9   => 'تسعة',
      10  => 'عشرة',
      11  => 'احد عشر',
      12  => 'اثنا عشر',
      13  => 'ثلاثة عشر',
      14  => 'أربعة عشر',
      15  => 'خمسة عشر',
      16  => 'ستة عشر',
      17  => 'سبعة عشر',
      18  => 'ثمانية عشر',
      19  => 'تسعة عشر',
      20  => 'عشرون',
      30  => 'ثلاثون',
      40  => 'أربعون',
      50  => 'خمسون',
      60  => 'ستون',
      70  => 'سبعون',
      80  => 'ثمانون',
      90  => 'تسعون',
      100 => 'مائة',
      200 => 'مئتان',
      2000 => 'ألفان'
    }.freeze

    AND_CONNECTOR = " و ".freeze
    DECIMAL_POINT_CONNECTOR = " فاصل ".freeze
    MINUS_WORD = "سالب ".freeze
  end
end


lib/number_humanizer/languages/english.rb
module NumberHumanizer::Languages
  class English < Base
    UNITS_DICTIONARY = {
      1 => '',
      2 => 'thousand',
      3 => 'million',
      4 => 'billion',
      5 => 'trillion',
      6 => 'quadrillion',
      7 => 'quintillion',
      8 => 'sextillion',
      9 => 'septillion',
      10 => 'octillion',
      11 => 'nonillion',
      12 => 'decillion',
      13 => 'undecillion',
      14 => 'duodecillion',
      15 => 'tredecillion',
      16 => 'quattuordecillion',
      17 => 'quindecillion',
      18 => 'sexdecillion',
      19 => 'septemdecillion',
      20 => 'octodecillion',
      21 => 'novemdecillion',
      22 => 'vigintillion'
    }.freeze

    NUMBERS_DICTIONARY = {
      0   => 'zero',
      1   => 'one',
      2   => 'two',
      3   => 'three',
      4   => 'four',
      5   => 'five',
      6   => 'six',
      7   => 'seven',
      8   => 'eight',
      9   => 'nine',
      10  => 'ten',
      11  => 'eleven',
      12  => 'twelve',
      13  => 'thirteen',
      14  => 'fourteen',
      15  => 'fifteen',
      16  => 'sixteen',
      17  => 'seventeen',
      18  => 'eighteen',
      19  => 'nineteen',
      20  => 'twenty',
      30  => 'thirty',
      40  => 'forty',
      50  => 'fifty',
      60  => 'sixty',
      70  => 'seventy',
      80  => 'eighty',
      90  => 'ninety',
      100 => 'hundred',
      200 => 'two hundred'
    }.freeze

    AND_CONNECTOR = " and ".freeze
    DECIMAL_POINT_CONNECTOR = " point ".freeze
    MINUS_WORD = "minus ".freeze

    private

    def process_words_for_hundreds(segment)
      word = "#{process_hundred_part(segment)}"
      word << "#{and_connector}#{process_words_for_tens(segment % 100)}" if segment % 100 > 0
      word
    end

    def process_words_for_tens(segment)
      return word_lookup(segment) if word_lookup(segment).present?
      "#{word_lookup(segment / 10 * 10)} #{word_lookup(segment % 10)}"
    end

    def process_hundred_part(segment)
      hundred_digit = segment / 100
      return '' if hundred_digit.zero?
      (1..2).cover?(hundred_digit) ? word_lookup(hundred_digit * 100) : "#{word_lookup(hundred_digit)} #{word_lookup(100)}"
    end
  end
end
